File: problem29/problem29.js
/* Problem 29
 * multiplication: takes a dynamic array of numbers, multiplies by an integer and returns the result in a new dynamic array e.g. multiplication([1, 2, 3], 10) returns [1, 2, 3, 0]
 * exponent: exponent(a, b) returns a^b e.g. exponent(2, 9) returns [5, 1, 2]
 * isEqual: checks whether two dynamic arrays have the same values e.g. isEqual([5, 1, 2], [5, 1, 2]) returns true
 * isIn: returns whether a dynamic array of dynamic arrays contains a dynamic array given
*/

var lowerBound = 2; var upperBound = 100; var nbOfDistinctTerms = 0;
var prevValues = [];
for (var p = lowerBound; p <= upperBound; p++) {
	for (var q = lowerBound; q <= upperBound; q++) {
		if (!isIn(prevValues, exponent(p, q))) {
			nbOfDistinctTerms++; prevValues.push(exponent(p, q));
		}
	}
}
console.log("The solution is ", nbOfDistinctTerms);

function multiplication(array, n) {
	var result = []; var carryOver = 0; var p = 0;
	for (var i = 0; i < array.length - 1; i++) {
		p = array[array.length - 1 - i] * n + carryOver; 
		result.push(p % 10); carryOver = Math.floor(p / 10); 
	}
	p = array[0] * n + carryOver;
	while (true) {
		result.push(p % 10);
		if (Math.floor(p / 10) == 0) break;
		else p = Math.floor(p / 10);
	}
	return result.reverse();
}

function exponent(a, b) {
	var result = [1];
	for (var i = 0; i < b; i++) {
		result = multiplication(result, a);
	}
	return result;
}

function isEqual(array1, array2) {
	if (array1.length != array2.length) return false;
	for (var i = 0; i < array1.length; i++) {
		if (array1[i] != array2[i]) return false;
	}
	return true;
}

function isIn(array1, array2) { // array 1 is an array of arrays
	for (var i = 0; i < array1.length; i++) {
		if (isEqual(array1[i], array2)) return true;
	}
	return false;
} 

File: problem29/Problem29.java
import java.util.*;
/* Problem 29
 * multiplication: takes a dynamic array of numbers, multiplies by an integer and returns the result in a new dynamic array e.g. multiplication([1, 2, 3], 10) returns [1, 2, 3, 0]
 * reverseInPlace: reverses a dynamic array
 * exponent: exponent(a, b) returns a^b e.g. exponent(2, 9) returns [5, 1, 2]
 * isEqual: checks whether two dynamic arrays have the same values e.g. isEqual([5, 1, 2], [5, 1, 2]) returns true
 * isIn: returns whether a dynamic array of dynamic arrays contains a dynamic array given
*/

public class Problem29 {
	public static void main(String[] arg) {
		int lowerBound = 2; int upperBound = 100; int nbOfDistinctTerms = 0;
		ArrayList<ArrayList<Integer>> prevValues = new ArrayList<ArrayList<Integer>>();
		for (int p = lowerBound; p <= upperBound; p++) {
			for (int q = lowerBound; q <= upperBound; q++) {
				if (!isIn(prevValues, exponent(p, q))) {
					nbOfDistinctTerms++; prevValues.add(exponent(p, q));
				}
			}
		}
		System.out.println("The solution is " + nbOfDistinctTerms);
	}
	
	public static ArrayList<Integer> multiplication(ArrayList<Integer> dynArray, int n) {
		ArrayList<Integer> result = new ArrayList<Integer>(); int carryOver = 0; int p = 0;
		for (int i = 0; i < dynArray.size() - 1; i++) { // there is a -1 as the first element will be treated as a separate case
			p = dynArray.get(dynArray.size() - 1 - i) * n + carryOver;
			result.add(p % 10); carryOver = p / 10;
		}
		p = dynArray.get(0) * n + carryOver; // treat the case of the first element
		while (true) {
			result.add(p % 10);
			if (p / 10 == 0) break;
			else p /= 10;
		}	
		reverseInPlace(result); return result;
	}

	public static void reverseInPlace(ArrayList<Integer> dynArray) {
		ArrayList<Integer> copy = new ArrayList<Integer>();
		for (int i = 0; i < dynArray.size(); i++) {
			copy.add(dynArray.get(i));
		}
		dynArray.clear();
		for (int i = 0; i < copy.size(); i++) {
			dynArray.add(copy.get(copy.size() - 1 - i));
		}
	}
	
	public static ArrayList<Integer> exponent(int a, int b) {
		ArrayList<Integer> result = new ArrayList<Integer>(); result.add(1);
		for (int i = 0; i < b; i++) {
			result = multiplication(result, a);
		}
		return result;
	}
	
	public static boolean isEqual(ArrayList<Integer> d1, ArrayList<Integer> d2) {
		if (d1.size() != d2.size()) return false;
		for (int i = 0; i < d1.size(); i++) {
			if (d1.get(i) != d2.get(i)) return false;
		}
		return true;	
	}
	
	public static boolean isIn(ArrayList<ArrayList<Integer>> d1, ArrayList<Integer> d2) {
		for (int i = 0; i < d1.size(); i++) {
			if (isEqual(d1.get(i), d2)) return true;
		}
		return false;
	}
}